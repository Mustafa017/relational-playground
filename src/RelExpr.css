.RelExpr .highlighted,
.RelExpr.highlighted {
  background-color: #ff3;
}

.RelExpr .clicked {
  background: #ff3;
  cursor: pointer;
}
