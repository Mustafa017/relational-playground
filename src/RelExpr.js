import React, { Component } from 'react';
import RelOp, { Projection, Rename, Selection } from './RelOp';

class RelExpr extends Component {
  constructor() {
    super();
    this.handleExprClick = this.handleExprClick.bind(this);
  }

  conditionToString(select, conds = []) {
    if (select.length === 0) { return conds; }

    const field = Object.keys(select[0])[0];
    const op = Object.keys(select[0][field])[0];
    const opMap = {
      '$gte': '>=',
      '$gt': '>=',
      '$lt': '<',
      '$lte': '<=',
      '$ne': '!=',
      '$eq': '=',
    };
    conds.push(field + ' ' + opMap[op] + ' ' + select[0][field][op]);

    return conds;
  }

  buildExpr(expr) {
    switch(Object.keys(expr)[0]) {
      case 'projection':
        return (
          <RelOp operator={<Projection project={expr.projection.arguments.project}/>}>
            <RelExpr expr={expr.projection.children[0]} changeExpr={this.props.changeExpr}/>
          </RelOp>
        );
      case 'selection':
        return (
          <RelOp operator={<Selection select={this.conditionToString(expr.selection.arguments.select)}/>}>
            <RelExpr expr={expr.selection.children[0]} changeExpr={this.props.changeExpr}/>
          </RelOp>
        );
      case 'rename':
        return (
          <RelOp operator={<Rename rename={expr.rename.arguments.rename}/>}>
            <RelExpr expr={expr.rename.children[0]} changeExpr={this.props.changeExpr}/>
          </RelOp>
        );
      case 'relation':
        return expr.relation;
      default:
        throw new Error('Invalid expression ' + JSON.stringify(expr) + '.');
    }
  }

  handleExprClick(e) {
    e.stopPropagation();
    this.props.changeExpr(this.props.expr);
  }

  render() {
    return <span onClick={this.handleExprClick}>{this.buildExpr(this.props.expr)}</span>;
  }
}

export default RelExpr;
