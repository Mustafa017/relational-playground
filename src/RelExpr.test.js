import React from 'react';
import renderer from 'react-test-renderer';

import RelExpr from './RelExpr';

/** @test {RelExpr} */
it('correctly renders a complex expression', () => {
  const expr = {
    rename: {
      arguments: {rename: {firstName: 'name'}},
      children: [
        {
          projection: {
            arguments: {project: ['firstName', 'lastName']},
            children: [
              {
                selection: {
                  arguments: {select: [{salary: {$gt: 130000}}]},
                  children: [{relation: 'Doctor'}],
                },
              },
            ],
          },
        },
      ],
    },
  };
  const tree = renderer.create(<RelExpr expr={expr} />).toJSON();
  expect(tree).toMatchSnapshot();
});

/** @test {RelExpr} */
it('produces an error for an invalid expression', () => {
  expect(() => {
    renderer.create(<RelExpr expr={{invalidExpr: 42}} />);
  }).toThrow();
});
