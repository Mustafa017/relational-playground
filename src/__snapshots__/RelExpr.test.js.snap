// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders a complex expression 1`] = `
<span
  className="RelExpr"
  onClick={[Function]}
  style={
    Object {
      "margin": ".4em",
    }
  }
>
  <span>
    <span
      className=""
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span>
        <span>
          ρ
          <sub>
            firstName/name
          </sub>
        </span>
        (
        <span
          className="RelExpr"
          onClick={[Function]}
          style={
            Object {
              "margin": ".4em",
            }
          }
        >
          <span>
            <span
              className=""
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <span>
                <span>
                  π
                  <sub>
                    firstName,lastName
                  </sub>
                </span>
                (
                <span
                  className="RelExpr"
                  onClick={[Function]}
                  style={
                    Object {
                      "margin": ".4em",
                    }
                  }
                >
                  <span>
                    <span
                      className=""
                      onMouseOut={[Function]}
                      onMouseOver={[Function]}
                    >
                      <span>
                        <span>
                          σ
                          <sub>
                            salary &gt; 130000
                          </sub>
                        </span>
                        (
                        <span
                          className="RelExpr"
                          style={
                            Object {
                              "margin": ".4em",
                            }
                          }
                        >
                          <span
                            className="Relation"
                            data-testid="span"
                          >
                            Doctor
                          </span>
                        </span>
                        )
                      </span>
                    </span>
                  </span>
                </span>
                )
              </span>
            </span>
          </span>
        </span>
        )
      </span>
    </span>
  </span>
</span>
`;
